Allow module extension usages to be isolated

fmeum opened the pull request:
If `isolate = True` is specified on `use_extension`, that particular usage will be isolated from all other usages, both in the same and other modules.

Module extensions can check whether they are isolated (e.g. in case they can only be used in this way) via `module_ctx.is_isolated`.

Wyverald commented on src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BazelDepGraphFunction.java:
```diff
@@ -248,7 +253,21 @@ private ImmutableBiMap<String, ModuleExtensionId> calculateUniqueNameForUsedExte
+          id.getNamespace()
+              .map(
+                  namespace ->
+                      nonEmptyRepoPart
```
using a String.format probably makes this easier to read

Wyverald commented on src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleExtensionId.java:
```diff
@@ -17,15 +17,28 @@
 
-  public static ModuleExtensionId create(Label bzlFileLabel, String extensionName) {
-    return new AutoValue_ModuleExtensionId(bzlFileLabel, extensionName);
+  public abstract Optional<Namespace> getNamespace();
```
IMO we don't need to generalize the concept of a "module extension ID namespace" until that's actually happened. For now, just calling this `isolatedToModuleKey` or some such is probably clearer.

Wyverald commented on src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileGlobals.java:
```diff
@@ -432,31 +432,42 @@ public void registerToolchains(boolean devDependency, Sequence<?> toolchainLabel
+        @Param(
+            name = "isolated",
+            doc =
+                "If true, this usage of the module extension will be isolated from all other "
```
we'll need to explain a bit better what "isolation" means here. question -- why "in particular those in other modules"? It creates isolation just as much for other usages in the current module, right?

fmeum replied:
I expanded the docs, please take another look. I would prefer to avoid the term "workspace" in favor of the less overloaded term "repository".

Wyverald replied:
I like the new docs!

aherrmann commented on src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileGlobals.java:
```diff
@@ -432,31 +432,42 @@ public void registerToolchains(boolean devDependency, Sequence<?> toolchainLabel
-      if (usageBuilder.extensionBzlFile.equals(extensionBzlFile)
-          && usageBuilder.extensionName.equals(extensionName)) {
-        return usageBuilder.getProxy(devDependency);
+    if (!isolated) {
+      for (ModuleExtensionUsageBuilder usageBuilder : extensionUsageBuilders) {
+        if (usageBuilder.extensionBzlFile.equals(extensionBzlFile)
+            && usageBuilder.extensionName.equals(extensionName)) {
+          return usageBuilder.getProxy(devDependency);
+        }
       }
     }
```
How do we prevent a non-isolated instance from getting mapped to a previously created isolated instance?

aherrmann replied:
Thank you, that looks good to me.

aherrmann reviewed (commented):
I like this idea. I think this would help with the distinction between globally unified and local packages in rules_nixpkgs, see [here](https://github.com/tweag/rules_nixpkgs/tree/0c1f8f5470c7f292b7620762e224f53d837929d3/design/bzlmod#unified-nix-package). The globally unified ones would be provided by a regular extension, and the local ones by an isolated extension.

fmeum commented:
I added tests as well as `module_ctx.extension_metadata` support. I also changed the naming scheme, see the comment in `BazelDepGraphFunction`.

@Wyverald https://github.com/bazelbuild/buildtools/pull/1171 is the corresponding change to buildozer.

fmeum commented:
Gson doesn't seem to handle `Optional<T>` out of the box. @SalmaSamy Do you happen to know what the best approach is to manually add support for it?

fmeum commented:
I ended up writing my own `TypeAdapterFactory`: https://github.com/bazelbuild/bazel/pull/18529/commits/23291be1f6edc4a8331ca624f86f7ce6bac6dc06

Wyverald commented on src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileGlobals.java:
```diff
@@ -432,31 +435,62 @@ public void registerToolchains(boolean devDependency, Sequence<?> toolchainLabel
 
     String extensionBzlFile = normalizeLabelString(rawExtensionBzlFile);
 
+    Optional<ModuleExtensionId.IsolationKey> isolationContext;
```
isolationKey?

Wyverald commented on src/test/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleExtensionResolutionTest.java:
```diff
@@ -611,6 +611,101 @@ public void multipleModules_ignoreDevDependency() throws Exception {
+        "    data_str += mod.name + '@' + mod.version + (' (root): ' if mod.is_root else ': ')",
+        "    for tag in mod.tags.tag:",
+        "      data_str += tag.data",
+        "      tag.expect_isolated == ctx.is_isolated or fail()",
```
this is very Perl-y :) maybe `if ...: fail()` instead?

Wyverald reviewed (approved):
nice!

fmeum commented:
Just in case you missed it: I changed the name of the parameter from `isolated` to `isolate` as I felt that matches better with `use_extension`. Let me know if you don't like this "bikeshe" :-)

fmeum commented:
@bazel-io flag

meteorcloudy commented:
@bazel-io fork 6.3.0
